enclave-runner/src/usercalls/mod.rs

```diff
@@ -347,36 +394,246 @@ impl EnclaveState {
             last_fd,
             exiting: AtomicBool::new(false),
             usercall_ext : usercall_ext,
+            work_sender: Mutex::new(work_sender)
         })
     }
 
+    fn create_worker_threads(num_of_worker_threads: usize,
+                             work_receiver: crossbeam::crossbeam_channel::Receiver<Work>,
+                             io_queue_send: mpsc::Sender<UsercallSendData>,
+                             enclave: Arc<EnclaveState>) {
+        for _ in 0..num_of_worker_threads {
+            let work_receiver = work_receiver.clone();
+            let io_queue_send = io_queue_send.clone();
+            let enclave = enclave.clone();
+
+            thread::spawn(move || {
+                loop {
+                    let work = work_receiver.recv();
+                    let work = match work {
+                        Err(crossbeam::channel::RecvError) => {
+                            break;
+                        },
+                        Ok(work) => work,
+                    };
+                    let res = work.do_work(&io_queue_send);
+                    if let Err(err) = res {
+                        match err.0 {
+                            (_, _, EnclaveEntry::ExecutableNonMain, _) => {
+
+                                let cmd = enclave.kind.as_command().unwrap();
+                                let mut cmddata = cmd.data.lock().unwrap();
+                                cmddata.running_secondary_threads -= 1;
```

Comment:
I don't think you need to do this accounting here, see my comment above about the kinds of errors you can get in `do_work`. Removing this then allows you to avoid passing `EnclaveState` to worker threads entirely.